Improve error message for Type checks

nikithamalgifb opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51842 Improve error message for Type checks**

Fixes issue #49511

Summary:
==========

Improves the error message for better type checking

Commit:
Improve error message for Type checks

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -81,7 +81,12 @@ struct Argument {
         actual_type,
         "'.\n",
-        inferred_type_hint);
+        inferred_type_hint,
```
This error message is indeed more actionable, but it doesn't provide additional information or addresses user confusion (difference between 2 and 2.0). Maybe be explicit in suggesting user to check for integer vs float literals?

nikithamalgifb replied:
@gmagogsfm : Can you please take a look at this fix?

gmagogsfm replied:
Optional:
There isn't a case where both inferred_type_hint and err_msg_for_int_float_literals are needed, you can replace both of them with a single variable, like "type_error_hint"

Commit:
Update on "Improve error message for Type checks"


Fixes issue #49511

Summary:
==========

Improves the error message for better type checking

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -72,6 +72,14 @@ struct Argument {
           name(),
           "' to be of type 'Tensor' ",
           "because it was not annotated with an explicit type.\n");
+    } else if(type()->repr_str() == "float" || type()->repr_str() == "int") {
```
Using `repr_str()` to tell types apart isn't as reliable as `kind()`.

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -72,6 +72,14 @@ struct Argument {
           "because it was not annotated with an explicit type.\n");
+    } else if(type()->repr_str() == "float" || type()->repr_str() == "int") {
+      if (actual_type == "int" || actual_type == "float") {
+      inferred_type_hint = c10::str(
```
Nit:
Consider using a different variable name because technically this error message is not about "inferred type".

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -72,6 +72,14 @@ struct Argument {
+    } else if(type()->repr_str() == "float" || type()->repr_str() == "int") {
+      if (actual_type == "int" || actual_type == "float") {
+      inferred_type_hint = c10::str(
+          "Did you forget to assign a '",
```
This is a lot better than last version, thanks! I think we can be even clearer here though to let users know about their misusing float/int literals, something like "Maybe you assigned a xxx literal (1.0) rather than yyy literal (1)"?

Commit:
Update on "Improve error message for Type checks"


Fixes issue #49511

Summary:
==========

Improves the error message for better type checking

[ghstack-poisoned]

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -66,12 +66,23 @@ struct Argument {
+    } else if(type()->kind() == FloatType::Kind || type()->kind() == IntType::Kind) {
+      std::string value_str = type()->kind() == FloatType::Kind ? "(1.0)" : "(1)";
+      std::string recieved_str = actual_type == "int" ? "(1)" : "(1.0)";
+      if (actual_type == "int" || actual_type == "float") {
```
IMHO, it might make sense to refactor formatTypeMismatchMsg to take in `TypePtr` instead of type strings for more reliable type comparison.

gmagogsfm commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -66,12 +66,23 @@ struct Argument {
+      err_msg_for_int_float_literals = c10::str(
+          "May be you assigned a '",
+           type()->repr_str(),
+          " literal ", value_str, " instead of '",
```
Please make it clear that (1 vs 1.0) are just examples, otherwise users will be confused about why there is a `1` in the first place when their literal might be 2 or 3.
